File: protocol/messages/Packet.kt
package com.example.universaldesktopapp.protocol

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.EOFException

enum class PacketType(val value: Byte) {
    Handshake(0x01),
    PairingRequest(0x02),
    PairingResponse(0x03),
    VideoFrame(0x04),
    MouseEvent(0x05),
    KeyEvent(0x06),
    Clipboard(0x07),
    FileMetadata(0x08),
    FileChunk(0x09),
    FileComplete(0x0A);

    companion object {
        fun fromValue(value: Byte): PacketType? = entries.find { it.value == value }
    }
}

class Packet(val type: PacketType, val payload: ByteArray) {
    val length: Int = payload.size

    fun serialize(outputStream: DataOutputStream) {
        outputStream.writeInt(length)
        outputStream.writeByte(type.value.toInt())
        outputStream.write(payload)
    }

    companion object {
        private const val MAX_PAYLOAD_SIZE = 32 * 1024 * 1024

        fun deserialize(inputStream: DataInputStream): Packet? {
            val length = try { inputStream.readInt() } catch (_: EOFException) { return null }
            require(length in 0..MAX_PAYLOAD_SIZE) { "Invalid packet length: $length" }
            val typeByte = inputStream.readByte()
            val type = requireNotNull(PacketType.fromValue(typeByte)) { "Unknown packet type: $typeByte" }
            return Packet(type, ByteArray(length).also(inputStream::readFully))
        }
    }
}

File: protocol/messages/Packet.cs
using System;
using System.IO;

namespace UniversalMobileDesktop.Protocol
{
    public enum PacketType : byte
    {
        Handshake = 0x01,
        PairingRequest = 0x02,
        PairingResponse = 0x03,
        VideoFrame = 0x04,
        MouseEvent = 0x05,
        KeyEvent = 0x06
    }

    public class Packet
    {
        public int Length { get; set; }
        public PacketType Type { get; set; }
        public byte[] Payload { get; set; }

        public Packet(PacketType type, byte[] payload)
        {
            Type = type;
            Payload = payload ?? Array.Empty<byte>();
            Length = Payload.Length;
        }

        public byte[] Serialize()
        {
            using var ms = new MemoryStream();
            using var writer = new BinaryWriter(ms);
            
            writer.Write(Length);
            writer.Write((byte)Type);
            writer.Write(Payload);
            
            return ms.ToArray();
        }

        public static Packet? Deserialize(BinaryReader reader)
        {
            try
            {
                int length = reader.ReadInt32();
                byte typeByte = reader.ReadByte();
                byte[] payload = reader.ReadBytes(length);
                
                return new Packet((PacketType)typeByte, payload);
            }
            catch (EndOfStreamException)
            {
                return null;
            }
        }
    }
}
